File: AI_Agents/frontend/api/fetch_ticket.js
export const config = { runtime: 'edge' };

function parseJiraIssue(issueData) {
    const fields = issueData.fields || {};
    return {
        summary: fields.summary || "",
        description: fields.description || "",
        issuetype: fields.issuetype?.name || "",
        status: fields.status?.name || ""
    };
}

export default async function handler(req) {
    if (req.method !== 'POST') return new Response('Method Not Allowed', { status: 405 });
    
    try {
        const payload = await req.json();
        const config = payload.source_config || {};
        const requestData = payload.request_data || {};
        
        let ticketId = requestData.ticket_id || '';
        if (ticketId && ticketId.includes('/')) ticketId = ticketId.split('/').pop().split('?')[0].trim();
        
        const provider = (config.provider || '').toLowerCase();
        
        if (provider === 'jira') {
            let domain = (config.domain_url || '').replace(/\/+$/, "");
            if (!domain.startsWith('http')) domain = `https://${domain}`;
            
            const email = config.email || '';
            const token = config.api_token || '';
            
            if (!domain || !email || !token || !ticketId) {
                return Response.json({ status: "error", message: "Missing required Jira fields" });
            }
            
            const url = `${domain}/rest/api/3/issue/${ticketId}?fields=summary,description,issuetype,status`;
            const auth = btoa(`${email}:${token}`);
            
            const resp = await fetch(url, {
                headers: { 'Authorization': `Basic ${auth}`, 'Accept': 'application/json' }
            });
            
            if (resp.ok) {
                const data = await resp.json();
                const parsed = parseJiraIssue(data);
                
                let context = `--- JIRA TICKET FETCHED [${ticketId}] ---\n\n`;
                context += `Subject: ${parsed.summary}\n`;
                context += `Type: ${parsed.issuetype} | Status: ${parsed.status}\n\n`;
                
                let descText = "Description:\n";
                const desc = parsed.description;
                
                if (typeof desc === 'object' && desc !== null) {
                    const extractText = (node) => {
                        let text = "";
                        if (node.type === "text") return node.text || "";
                        if (node.content) {
                            node.content.forEach(child => text += extractText(child) + " ");
                        }
                        if (node.type === "paragraph") text += "\n";
                        return text;
                    };
                    descText += extractText(desc);
                } else {
                    descText += String(desc || "");
                }
                
                context += descText;
                return Response.json({ status: "success", ticket_id: ticketId, ticket_context: context });
            } else {
                 let errText = await resp.text();
                 try { const j = JSON.parse(errText); if (j.errorMessages) errText = j.errorMessages[0]; } catch(e){}
                 if (resp.status === 401) errText = "Invalid API Token or Email. Ensure you use an API Token, NOT your password.";
                 return Response.json({ status: "error", message: `HTTP ${resp.status}: ${errText}` });
            }
        }
        return Response.json({ status: "error", message: "Unsupported provider" });
    } catch (e) {
        return Response.json({ status: "error", message: e.message });
    }
}

File: AI_Agents/tools/fetch_ticket.py
import sys
import json
import requests
import traceback

def parse_jira_issue(issue_data):
    fields = issue_data.get("fields", {})
    return {
        "summary": fields.get("summary", ""),
        "description": fields.get("description", ""),
        "issuetype": fields.get("issuetype", {}).get("name", ""),
        "status": fields.get("status", {}).get("name", "")
    }

def fetch_source_ticket(config, ticket_id):
    # Extract the key if a full URL was pasted (e.g., https://.../browse/KAN-4)
    if ticket_id and "/" in ticket_id:
        ticket_id = ticket_id.split("/")[-1].split("?")[0].strip()
        
    provider = config.get("provider", "").lower()
    if provider == "jira":
        domain = config.get("domain_url", "").strip("/")
        if not domain.startswith("http"):
             domain = f"https://{domain}"
             
        email = config.get("email", "")
        token = config.get("api_token", "")
        
        if not domain or not email or not token or not ticket_id:
            return {"status": "error", "message": "Missing required fields to fetch Jira ticket"}
            
        url = f"{domain}/rest/api/3/issue/{ticket_id}"
        headers = {"Accept": "application/json"}
        
        try:
            resp = requests.get(url, headers=headers, auth=(email, token), params={"fields": "summary,description,issuetype,status"}, timeout=15)
            if resp.status_code == 200:
                data = resp.json()
                parsed = parse_jira_issue(data)
                
                context = f"--- JIRA TICKET FETCHED [{ticket_id}] ---\n\n"
                context += f"Subject: {parsed['summary']}\n"
                context += f"Type: {parsed['issuetype']} | Status: {parsed['status']}\n\n"
                
                desc = parsed['description']
                desc_text = "Description:\n"
                if isinstance(desc, dict):
                    # Basic ADF extraction
                    def extract_text(node):
                        text = ""
                        if node.get("type") == "text":
                            return node.get("text", "")
                        for child in node.get("content", []):
                            text += extract_text(child) + " "
                        if node.get("type") == "paragraph":
                            text += "\n"
                        return text
                    desc_text += extract_text(desc)
                else:
                    desc_text += str(desc)
                    
                context += desc_text
                
                return {"status": "success", "ticket_id": ticket_id, "ticket_context": context}
            else:
                 try:
                     err_text = resp.json().get("errorMessages", [resp.text])[0]
                 except:
                     err_text = resp.text
                 if resp.status_code == 401:
                     err_text = "Invalid API Token or Email. Ensure you are using an Atlassian API Token generated from id.atlassian.com, NOT your login password."
                 return {"status": "error", "message": f"HTTP {resp.status_code}: {err_text}"}
        except Exception as e:
            return {"status": "error", "message": f"Failed to fetch ticket: {str(e)}"}
            
    return {"status": "error", "message": f"Unsupported or unspecified provider: '{provider}'."}

def main():
    if len(sys.argv) < 2:
        print(json.dumps({"status": "error", "message": "Missing JSON payload"}))
        sys.exit(1)
        
    try:
        payload = json.loads(sys.argv[1])
        source_config = payload.get("source_config", {})
        request_data = payload.get("request_data", {})
        
        ticket_id = request_data.get("ticket_id", "")
        result = fetch_source_ticket(source_config, ticket_id)
        print(json.dumps(result))
    except Exception as e:
        print(json.dumps({"status": "error", "message": f"Internal Fetch Error: {str(e)}", "trace": traceback.format_exc()}))

if __name__ == "__main__":
    main()
